src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSystemHookAction.java

```diff
@@ -66,10 +70,25 @@ public HttpResponse doPost(StaplerRequest request) throws GitLabApiException {
             return HttpResponses.error(HttpServletResponse.SC_BAD_REQUEST,
                 "Expecting a GitLab event, missing expected X-Gitlab-Event header");
         }
+        String secretToken = request.getHeader("X-Gitlab-Token");
+        if(!isValidToken(secretToken)) {
+            return HttpResponses.error(HttpServletResponse.SC_UNAUTHORIZED,
+                "Expecting a valid secret token");
+        }
         String origin = SCMEvent.originOf(request);
         SystemHookManager systemHookManager = new SystemHookManager();
         systemHookManager.addListener(new GitLabSystemHookListener(origin));
         systemHookManager.handleEvent(request);
         return HttpResponses.ok(); // TODO find a better response
     }
+
+    private boolean isValidToken(String secretToken) {
+        List<GitLabServer> servers = GitLabServers.get().getServers();
+        for(GitLabServer server: servers) {
+            if(server.getSecretToken().equals(secretToken)) {
```

Comment:
Might not be the correct way as there are potential null pointers lurking.
```suggestion
            if(server.getSecretToken().getPlainValue().equals(secretToken)) {
```

Reply:
Adding a try catch should solve this?